README.md
# Tox
[![Hex.pm: version](https://img.shields.io/hexpm/v/tox.svg?style=flat-square)](https://hex.pm/packages/tox)
[![GitHub: CI status](https://img.shields.io/github/workflow/status/hrzndhrn/tox/CI?style=flat-square)](https://github.com/hrzndhrn/tox/actions)
[![Coveralls: coverage](https://img.shields.io/coveralls/github/hrzndhrn/tox?style=flat-square)](https://coveralls.io/github/hrzndhrn/tox)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg?style=flat-square)](https://github.com/hrzndhrn/tox/blob/main/LICENSE.md)

Some structs and functions to work with dates, times, periods, and intervals.

`Tox` needs Elixir 1.8 or greater. The library works with any
[`Calendar.time_zone_database`](https://hexdocs.pm/elixir/Calendar.html#t:time_zone_database/0)
and any [`Calendar`](https://hexdocs.pm/elixir/Calendar.html#content).

The tests and examples are using the time zone database from
[`TimeZoneInfo`](https://github.com/hrzndhrn/time_zone_info)
and the calendars
[`Calendar.ISO`](https://hexdocs.pm/elixir/Calendar.ISO.html#content),
[`Cldr.Calendar.Coptic`](https://github.com/elixir-cldr/cldr_calendars_coptic)
and
[`Cldr.Calendar.Ethiopic`](https://github.com/elixir-cldr/cldr_calendars_ethiopic).

## Installation

The package can be installed by adding `tox` to your list of dependencies in
mix.exs. Fox time zone support you can install `time_zone_info`, `tz`, or
`tzdata`.

```elixir
def deps do
  [
    {:tox, "~> 0.2"}
    # add one of the following packages for time zone support
    # {:time_zone_info, "~> 0.5"}
    # {:tz, "~> 0.10"}
    # {:tzdata, "~> 1.0"}
  ]
end
```

## Notes

`Tox` assumes that a week consists of 7 days. This assumption is not true for all
calendars.

The functions `Tox.DateTime.week/1`, `Tox.NaiveDateTime.week/1`, and
`Tox.Date.week/1` are just defined for `Calendar.ISO`.
This functions are returning a tuple of the year and the week number.

## Usage

Some examples how to use `Tox`.  For more information and examples, see
[`API reference`](https://hexdocs.pm/tox/api-reference.html#content).

before and after:
```elixir
iex> datetime = DateTime.from_naive!(~N[2020-03-10 12:00:00], "Europe/Berlin")
iex> Tox.DateTime.before?(datetime, Tox.DateTime.shift(datetime, day: 1))
true
iex> Tox.DateTime.before_or_equal?(datetime, datetime)
true
iex> Tox.DateTime.after?(datetime, Tox.DateTime.shift(datetime, day: 1))
false
```

shifting back and forward:
```elixir
iex> datetime = DateTime.from_naive!(~N[2020-01-01 12:00:00], "Europe/Oslo")
iex> Tox.DateTime.shift(datetime, year: 2, month: -3, day: 2, hour: -24, minute: 10)
#DateTime<2021-10-02 12:10:00+02:00 CEST Europe/Oslo>

iex> naive_datetime = ~N[2020-01-01 12:00:00]
iex> Tox.NaiveDateTime.shift(naive_datetime, year: 2, month: -3, day: 2, hour: -24, minute: 10)
~N[2021-10-02 12:10:00+02:00]

iex> Tox.Date.shift(~D[2020-01-01], year: 2, month: -3, day: 2)
~D[2021-10-03]

iex> Tox.Time.shift(~T[00:00:00], hour: 2, minute: -10, microsecond: 99)
~T[01:50:00.000099]

iex> {:ambiguous, datetime1, datetime2} =
...>   DateTime.from_naive(~N[2020-10-25 02:30:00], "Europe/Oslo")
iex> Tox.DateTime.shift(datetime1, hour: 1) == datetime2
true

# falling into a gap
iex> datetime = DateTime.from_naive!(~N[2020-03-29 01:30:00], "Europe/Paris")
#DateTime<2020-03-29 01:30:00+01:00 CET Europe/Paris>
iex> Tox.DateTime.shift(datetime, minute: 40)
#DateTime<2020-03-29 03:10:00+02:00 CEST Europe/Paris>
```

period and interval:
```elixir
iex> now = DateTime.from_naive!(~N[2020-07-12 21:33:43], "America/New_York")
iex> period = Tox.Period.new!(month: 1)
#Tox.Period<P1M>
iex> interval = Tox.Interval.new!(Tox.DateTime.beginning_of_month(now), period)
#Tox.Interval<[2020-07-01T00:00:00-04:00/P1M[>
iex> Tox.Interval.contains?(interval, now)
true
iex> Tox.Interval.since_start(interval, now, :millisecond)
{:ok, 1028023000}
iex> Tox.Interval.until_ending(interval, now, :millisecond)
{:ok, 1650377000}
iex> Tox.Interval.until_ending(interval, Tox.DateTime.shift(now, month: 1), :millisecond)
:error
```

beginning and end of
```elixir
iex> ~N[2020-07-12 11:12:13]
...> |> DateTime.from_naive!("America/Winnipeg")
...> |> Tox.DateTime.beginning_of_year()
#DateTime<2020-01-01 00:00:00-06:00 CST America/Winnipeg>

iex> ~N[2020-07-12 11:12:13]
...> |> DateTime.from_naive!("Europe/Berlin")
...> |> Tox.DateTime.end_of_year()
#DateTime<2020-12-31 23:59:59.999999+01:00 CET Europe/Berlin>

# In 1994 the year ended one day earlier in the time zone Pacific/Kirimati.
iex> ~N[1994-07-12 11:12:13]
...> |> DateTime.from_naive!("Pacific/Kiritimati")
...> |> Tox.DateTime.end_of_year()
#DateTime<1994-12-30 23:59:59.999999-10:00 -10 Pacific/Kiritimati>

iex> Tox.NaiveDateTime.beginning_of_day(~N[2020-01-01 12:00:00])
~N[2020-01-01 00:00:00]

iex> Tox.NaiveDateTime.end_of_day(~N[2020-01-01 12:00:00])
~N[2020-01-01 23:59:59.999999]

iex> Tox.Date.beginning_of_week(~D[2020-01-10])
~D[2020-01-06]

iex> Tox.Date.end_of_week(~D[2020-01-10])
~D[2020-01-12]
```


CHANGELOG.md
# Changelog

## 0.2.3 - 2021/10/23

- Fix `beginning_of_week` functions.

## 0.2.2 - 2021/09/24

- Corrections in the documentation.
- Fix `beginning_of_week` functions.

## 0.2.1 - 2021/08/24

- Changes to handle `Calendar.day_of_week/4`.

## 0.2.0 - 2020/07/21

- Rename `add` functions to `shift`.
- Add `Tox.shift/2`.

## 0.1.2 - 2020/07/20

- The very first version.


lib/tox/date.ex
defmodule Tox.Date do
  @moduledoc """
  A set of functions to work with `Date`.
  """

  @doc """
  Shifts the `date` by the given `duration`.


  The `durations` is a keyword list of one or more durations of the type
  `Tox.duration` e.g. `[year: 1, month: 5, day: 500]`. All values will be
  shifted from the largest to the smallest unit.

  ## Examples

      iex> date = ~D[1980-11-01]
      iex> Tox.Date.shift(date, year: 2)
      ~D[1982-11-01]
      iex> Tox.Date.shift(date, year: -2, month: 1, day: 40)
      ~D[1979-01-10]
      # time units will be ignored
      iex> Tox.Date.shift(date, hour: 100, minute: 10, second: 10)
      ~D[1980-11-01]

  Adding a month at the end of the month can update the day too.

      iex> Tox.Date.shift(~D[2000-01-31], month: 1)
      ~D[2000-02-29]

  For that reason it is important to know that all values will be shifted from the
  largest to the smallest unit.

      iex> date = ~D[2000-01-30]
      iex> Tox.Date.shift(date, month: 1, day: 1)
      ~D[2000-03-01]
      iex> date |> Tox.Date.shift(month: 1) |> Tox.Date.shift(day: 1)
      ~D[2000-03-01]
      iex> date |> Tox.Date.shift(day: 1) |> Tox.Date.shift(month: 1)
      ~D[2000-02-29]

  Using `shift/2` with a different calendar.

      iex> ~D[2000-12-30]
      ...> |> Date.convert!(Cldr.Calendar.Coptic)
      ...> |> Tox.Date.shift(day: 3)
      %Date{year: 1717, month: 4, day: 24, calendar: Cldr.Calendar.Coptic}

  """
  @spec shift(Calendar.date(), [Tox.duration()]) :: Date.t()
  def shift(date, durations) do
    date
    |> shift_years(Keyword.get(durations, :year, 0))
    |> shift_months(Keyword.get(durations, :month, 0))
    |> Date.add(
      Keyword.get(durations, :day, 0) + Keyword.get(durations, :week, 0) * Tox.days_per_week()
    )
  end

  @doc """
  Returns an `{year, week}` representing the ISO week number for the specified
  date.

  This function is just defined for dates with `Calendar.ISO`.

  ## Example

      iex> Tox.Date.week(~D[2017-01-01])
      {2016, 52}
      iex> Tox.Date.week(~D[2020-01-01])
      {2020, 1}
      iex> Tox.Date.week(~D[2019-12-31])
      {2020, 1}
      iex> ~D[2020-06-04]
      ...> |> Date.convert(Cldr.Calendar.Ethiopic)
      ...> |> Tox.Date.week()
      ** (FunctionClauseError) no function clause matching in Tox.Date.week/1

  """
  @spec week(Calendar.date()) :: {Calendar.year(), non_neg_integer}
  def week(%{calendar: Calendar.ISO} = date), do: Tox.week(date)

  @doc """
  Returns true if `date1` occurs after `date2`.

  ## Examples

      iex> Tox.Date.after?(~D[2020-06-14], ~D[2020-06-22])
      false

      iex> Tox.Date.after?(~D[2020-07-14], ~D[2020-06-22])
      true

      iex> Tox.Date.after?(~D[2020-01-01], ~D[2020-01-01])
      false

      iex> Tox.Date.after?(
      ...>   Date.convert!(~D[2000-01-22], Cldr.Calendar.Coptic),
      ...>   Date.convert!(~D[2000-01-01], Cldr.Calendar.Coptic)
      ...> )
      true

  """
  defmacro after?(date1, date2) do
    quote do
      Date.compare(unquote(date1), unquote(date2)) == :gt
    end
  end

  @doc """
  Returns true if `date1` occurs after `date2` or both dates are equal.

  ## Examples

      iex> Tox.Date.after_or_equal?(~D[2020-06-14], ~D[2020-06-22])
      false

      iex> Tox.Date.after_or_equal?(~D[2020-07-14], ~D[2020-06-22])
      true

      iex> Tox.Date.after_or_equal?(~D[2020-01-01], ~D[2020-01-01])
      true

      iex> Tox.Date.after_or_equal?(
      ...>   Date.convert!(~D[2000-01-22], Cldr.Calendar.Ethiopic),
      ...>   Date.convert!(~D[2000-01-01], Cldr.Calendar.Ethiopic)
      ...> )
      true

  """
  defmacro after_or_equal?(date1, date2) do
    quote do
      Date.compare(unquote(date1), unquote(date2)) in [:gt, :eq]
    end
  end

  @doc """
  Returns true if both datets are equal.

  ## Examples

      iex> Tox.Date.equal?(~D[2020-07-14], ~D[2020-06-22])
      false

      iex> Tox.Date.equal?(~D[2020-01-01], ~D[2020-01-01])
      true

      iex> ethiopic = Date.convert!(~D[2000-01-01], Cldr.Calendar.Ethiopic)
      %Date{year: 1992, month: 4, day: 22, calendar: Cldr.Calendar.Ethiopic}
      iex> coptic = Date.convert!(~D[2000-01-01], Cldr.Calendar.Coptic)
      %Date{year: 1716, month: 4, day: 22, calendar: Cldr.Calendar.Coptic}
      iex> Tox.Date.equal?(ethiopic, coptic)
      true

  """
  defmacro equal?(date1, date2) do
    quote do
      Date.compare(unquote(date1), unquote(date2)) == :eq
    end
  end

  @doc """
  Returns true if `date1` occurs before `date2`.

  ## Examples

      iex> Tox.Date.before?(~D[2020-06-14], ~D[2020-06-22])
      true

      iex> Tox.Date.before?(~D[2020-07-14], ~D[2020-06-22])
      false

      iex> Tox.Date.before?(~D[2020-01-01], ~D[2020-01-01])
      false

      iex> Tox.Date.before?(
      ...>   Date.convert!(~D[2000-01-22], Cldr.Calendar.Ethiopic),
      ...>   Date.convert!(~D[2000-06-01], Cldr.Calendar.Ethiopic)
      ...> )
      true

  """
  defmacro before?(date1, date2) do
    quote do
      Date.compare(unquote(date1), unquote(date2)) == :lt
    end
  end

  @doc """
  Returns true if `date1` occurs before `date2` or both dates are equal.

  ## Examples

      iex> Tox.Date.before_or_equal?(~D[2020-06-14], ~D[2020-06-22])
      true

      iex> Tox.Date.before_or_equal?(~D[2020-07-14], ~D[2020-06-22])
      false

      iex> Tox.Date.before_or_equal?(~D[2020-01-01], ~D[2020-01-01])
      true

      iex> Tox.Date.before_or_equal?(
      ...>   Date.convert!(~D[2000-01-22], Cldr.Calendar.Ethiopic),
      ...>   Date.convert!(~D[2000-06-01], Cldr.Calendar.Ethiopic)
      ...> )
      true

  """
  defmacro before_or_equal?(date1, date2) do
    quote do
      Date.compare(unquote(date1), unquote(date2)) in [:lt, :eq]
    end
  end

  @doc """
  Returns a boolean indicating whether `date` occurs between `from` and `to`.
  The optional `boundaries` specifies whether `from` and `to` are included or
  not. The possible value for `boundaries` are:

  * `:open`: `from` and `to` are excluded
  * `:closed`: `from` and `to` are included
  * `:left_open`: `from` is excluded and `to` is included
  * `:right_open`: `from` is included and `to` is excluded

  ## Examples

      iex> from = ~D[2020-02-01]
      iex> to   = ~D[2020-03-01]
      iex> Tox.Date.between?(~D[2020-01-01], from, to)
      false
      iex> Tox.Date.between?(~D[2020-02-05], from, to)
      true
      iex> Tox.Date.between?(~D[2020-03-05], from, to)
      false
      iex> Tox.Date.between?(~D[2020-02-01], from, to)
      true
      iex> Tox.Date.between?(~D[2020-03-01], from, to)
      false
      iex> Tox.Date.between?(~D[2020-02-01], from, to, :open)
      false
      iex> Tox.Date.between?(~D[2020-03-01], from, to, :open)
      false
      iex> Tox.Date.between?(~D[2020-02-01], from, to, :closed)
      true
      iex> Tox.Date.between?(~D[2020-03-01], from, to, :closed)
      true
      iex> Tox.Date.between?(~D[2020-02-01], from, to, :left_open)
      false
      iex> Tox.Date.between?(~D[2020-03-01], from, to, :left_open)
      true
      iex> Tox.Date.between?(~D[2000-01-01], to, from)
      ** (ArgumentError) from is equal or greater as to

  """
  @spec between?(Calendar.date(), Calendar.date(), Calendar.date(), Tox.boundaries()) ::
          boolean()
  def between?(date, from, to, boundaries \\ :right_open)
      when boundaries in [:closed, :left_open, :right_open, :open] do
    if Date.compare(from, to) in [:gt, :eq],
      do: raise(ArgumentError, "from is equal or greater as to")

    case {Date.compare(date, from), Date.compare(date, to), boundaries} do
      {:lt, _, _} -> false
      {_, :gt, _} -> false
      {:eq, _, :closed} -> true
      {:eq, _, :right_open} -> true
      {_, :eq, :closed} -> true
      {_, :eq, :left_open} -> true
      {:gt, :lt, _} -> true
      {_, _, _} -> false
    end
  end

  @doc """
  Returns a date representing the start of the year.

  ## Examples

      iex> Tox.Date.beginning_of_year(~D[2020-11-11])
      ~D[2020-01-01]

  """
  @spec beginning_of_year(Calendar.date()) :: Calendar.date()
  def beginning_of_year(date), do: %{date | month: 1, day: 1}

  @doc """
  Returns a date representing the start of the month.

  ## Examples

      iex> Tox.Date.beginning_of_month(~D[2020-11-11])
      ~D[2020-11-01]

  """
  @spec beginning_of_month(Calendar.date()) :: Calendar.date()
  def beginning_of_month(date), do: %{date | day: 1}

  @doc """
  Returns a date representing the start of the week.

  ## Examples

      iex> Tox.Date.beginning_of_week(~D[2020-11-13])
      ~D[2020-11-09]

  """
  @spec beginning_of_week(Calendar.date()) :: Calendar.date()
  def beginning_of_week(date) do
    shift(date, day: Tox.Calendar.beginning_of_week(date))
  end

  @doc """
  Returns a date representing the end of the year.
  If the date cannot be determined, `{:error, reason}` is returned.

  ## Examples

      iex> Tox.Date.end_of_year(~D[2020-11-11])
      ~D[2020-12-31]

      iex> ~D[2020-11-11]
      iex> |> Date.convert!(Cldr.Calendar.Coptic)
      iex> |> Tox.Date.end_of_year()
      %Date{year: 1737, month: 13, day: 5, calendar: Cldr.Calendar.Coptic}

  """
  @spec end_of_year(Calendar.date()) :: Calendar.date()
  def end_of_year(%{calendar: calendar, year: year} = date) do
    month = calendar.months_in_year(year)
    day = calendar.days_in_month(year, month)
    %{date | month: month, day: day}
  end

  @doc """
  Returns a date representing the end of the month.

  ## Examples

      iex> Tox.Date.end_of_month(~D[2020-11-11])
      ~D[2020-11-30]

      iex> ~D[2020-12-31]
      ...> |> Date.convert!(Cldr.Calendar.Coptic)
      ...> |> Tox.Date.shift(day: 1)
      ...> |> Tox.Date.end_of_month()
      %Date{year: 1737, month: 4, day: 30, calendar: Cldr.Calendar.Coptic}

  """
  @spec end_of_month(Calendar.date()) :: Calendar.date()
  def end_of_month(%{calendar: calendar, year: year, month: month} = date) do
    day = calendar.days_in_month(year, month)
    %{date | day: day}
  end

  @doc """
  Returns a date representing the end of the week.

  ## Examples

      iex> Tox.Date.end_of_week(~D[2020-11-11])
      ~D[2020-11-15]

      iex> ~D[2020-11-11]
      ...> |> Date.convert!(Cldr.Calendar.Ethiopic)
      ...> |> Tox.Date.end_of_week()
      %Date{year: 2013, month: 3, day: 6, calendar: Cldr.Calendar.Ethiopic}

  """
  @spec end_of_week(Calendar.date()) :: Calendar.date()
  def end_of_week(%{calendar: calendar, year: year, month: month, day: day} = date) do
    day = Tox.days_per_week() - Tox.day_of_week(calendar, year, month, day)
    shift(date, day: day)
  end

  ## Helpers

  defp shift_years(date, 0), do: date

  defp shift_years(
         %{calendar: calendar, year: year, month: month, day: day} = date,
         years
       ) do
    updated_year = year + years
    updated_day = update_day(updated_year, month, day, calendar)

    %{date | year: updated_year, day: updated_day}
  end

  defp shift_months(date, 0), do: date

  defp shift_months(
         %{calendar: calendar, month: month, year: year, day: day} = date,
         months
       ) do
    {updated_year, updated_month} = shift_months(months, year, month, calendar)
    updated_day = update_day(updated_year, updated_month, day, calendar)

    %{date | year: updated_year, month: updated_month, day: updated_day}
  end

  defp shift_months(months, year, month, calendar) do
    months_per_year = calendar.months_in_year(year)

    updated_year = year + div(months, months_per_year)
    updated_month = month + rem(months, months_per_year)

    cond do
      updated_month <= 0 ->
        {updated_year - 1, months_per_year + updated_month}

      updated_month > months_per_year ->
        {updated_year + 1, updated_month - months_per_year}

      true ->
        {updated_year, updated_month}
    end
  end

  defp update_day(year, month, day, calendar), do: min(day, calendar.days_in_month(year, month))
end
